tensorizer/stream_io.py

```diff
@@ -417,6 +422,292 @@ def seek(self, position, whence=SEEK_SET):
+            self._curr_buffer_view = self._curr_buffer_view[num_bytes:]
+            return num_bytes
+
+        self._redis_tcp_mutex.acquire(True)
+
+        # We have an aligned read, so we can use GET.
+        if position == curr_key_pos:
+            command = f"GET {self._keys[curr_key].decode('utf-8')}"
+        else:
+            command = (
+                f"GETRANGE {self._keys[curr_key].decode('utf-8')}"
+                f" {begin_idx} {begin_idx + read_sz - 1}"
+            )
+
+        begin_timer = time.monotonic()
+        self._redis_tcp.send(command.encode("utf-8") + b"\r\n")
+        value_sz = self._read_sz()
+        tcp_read_sz = min(value_sz, read_sz)
+        if tcp_read_sz == 0:
+            self._redis_tcp_mutex.release()
+            return 0
+        num_bytes = self._redis_tcp.recv_into(
+            ba, tcp_read_sz, socket.MSG_WAITALL
+        )
+
+        end_timer = time.monotonic()
+        self.response_latencies.append(end_timer - begin_timer)
+
+        # We performed a partial read of the entire key, so we need to
+        # store the remainder in the buffer.
+        if value_sz > read_sz:
+            tcp_read_sz = value_sz - num_bytes
+            if tcp_read_sz > len(self._curr_buffer):
+                self._curr_buffer = bytearray(tcp_read_sz)
+            self._curr_buffer_view = memoryview(self._curr_buffer)[:tcp_read_sz]
+            self._redis_tcp.recv_into(
+                self._curr_buffer_view, tcp_read_sz, socket.MSG_WAITALL
+            )
+            self._curr_buffer_idx = position + num_bytes
+
+        # Read the trailing \r\n and discard.
+        self._redis_tcp.recv(2)
+
+        self._redis_tcp_mutex.release()
+        return num_bytes
+
+    def _read_sz(self) -> int:
+        # Loop until we get a \r\n
+        sz_resp = b""
+        while True:
+            b = self._redis_tcp.recv(1)
+            sz_resp += b
+            if sz_resp[-2:] == b"\r\n":
+                break
+        sz_str = sz_resp.decode("utf-8").strip()[1:]
+        if sz_str == "-1":
+            raise IOError("Key not found")
+        return int(sz_str)
+
+    def _read_until(
+        self, goal_position: int, ba: Optional[bytearray] = None
+    ) -> Union[bytes, int]:
+        try:
+            if ba is None:
+                rq_sz = goal_position - self._curr
+                mv = memoryview(bytearray(rq_sz))
+            else:
+                rq_sz = len(ba)
+                mv = memoryview(ba)
+            orig_mv = mv
+            left = rq_sz
+            while left > 0:
+                num_bytes = self._read_from(self._curr, left, mv)
+                if num_bytes == 0:
+                    break
+                left -= num_bytes
+                self._curr += num_bytes
+                if left == 0:
+                    break
+                mv = mv[num_bytes:]
+            self.bytes_read += rq_sz - left
+            if ba is None:
+                return orig_mv.tobytes()
```

Comment:
This requires an extra `memcpy` of all of the data from the mutable `bytearray` to the new immutable `bytes` object. If it is permissible to return a `bytearray` from this method, returning `orig_mv.obj` instead would be more performant.

Reply:
Didn't know about `obj`, as the type was opaque to me on inspection. Cool!